### ISSUE TITLE
Understanding of num_classes in code

### ISSUE THREAD

**Issue by weicheng113** [IDEA-Research/detrex#174]
Dear Authors, thanks for sharing high performance models.

I am reading through the DINO model code and get some questions below. Could you please help me?

1. Is the num_classes = actual_num_classes + 1(background class) in detrex?
https://github.com/IDEA-Research/detrex/blob/697f5e9dafab6ea1769ec2ea1e0b65351273aa32/projects/dino/modeling/dino.py#L101

The reason I am asking is because background class should also need a label, as I can see in DINO repo.

https://github.com/IDEA-Research/DINO/blob/66d7173cc4167934381a898b07c08507bdd96b63/models/dino/dino.py#L81
`self.label_enc = nn.Embedding(dn_labelbook_size + 1, hidden_dim)`

2. src_logits.shape[batch_size, n_dn_queries=900, num_classes] and target_classes.shape is [batch_size, n_dn_queries=900]. Does num_classes include background class, which means the last logit value is for background class? Is label_id for classes starting from "1" (I can see category_id is starting from 1 in coco dataset)? 
https://github.com/IDEA-Research/detrex/blob/697f5e9dafab6ea1769ec2ea1e0b65351273aa32/detrex/modeling/criterion/criterion.py#L112

https://github.com/IDEA-Research/detrex/blob/697f5e9dafab6ea1769ec2ea1e0b65351273aa32/detrex/modeling/criterion/criterion.py#L122

If num_classes already includes background class, then +1 in this line is not needed(but cross-entropy loss is not in use, so it does not matter.)?
https://github.com/IDEA-Research/detrex/blob/697f5e9dafab6ea1769ec2ea1e0b65351273aa32/detrex/modeling/criterion/criterion.py#L103

I was trying to apply DINO model in my custom dataset. So far it can train, but the performance is not so good. I think I might misunderstand `num_classes`.

=======UPDATE======

https://github.com/IDEA-Research/detrex/blob/697f5e9dafab6ea1769ec2ea1e0b65351273aa32/detrex/modeling/criterion/criterion.py#L141

I went through the code second time. It looks like for Focal Loss, num_classes only needs to = actual_num_classes(without +1 for background class). For example, there is a dataset with 2 classes: 0, 1. The logits for each prediction only needs 2 numbers, e.g., [0.0145, 0.0111]. If it is mapped to background class '2', onehot encoding will be [0, 0, 1]. We can cut off the last digit of onehot so that [0.0145, 0.0111] is comparing with [0, 0].

So with Focal Loss, we only need to set num_classes=actual_num_classes(without +1 for background class) for all, including the following two locations.

https://github.com/IDEA-Research/detrex/blob/697f5e9dafab6ea1769ec2ea1e0b65351273aa32/projects/dino/modeling/dino.py#L92

https://github.com/IDEA-Research/detrex/blob/697f5e9dafab6ea1769ec2ea1e0b65351273aa32/projects/dino/modeling/dino.py#L101

Background class concept is only limited inside SetCriterion class when trying to produce one-hot encoding and the last digit will be cut off, becoming all zeros for background class.

https://github.com/IDEA-Research/detrex/blob/697f5e9dafab6ea1769ec2ea1e0b65351273aa32/detrex/modeling/criterion/criterion.py#L116

Is my understanding correct?

Thanks,
Cheng

**Comment by weicheng113** [comment]
@HaoZhang534 Thanks for your time and help. 

I meant the loss was decreasing as expected. Take below loss output for example.
```python
loss=7.69, lr=0.0001, loss_class=0.00258, loss_bbox=0.0155, loss_giou=0.222...]  
```

And I loaded the trained model and tested on images, it gave fairly good predictions on many images. Therefore, I think the model was learning with the custom dataset.

I am not sure where I mis-configured the evaluator part, as it depends on annotation file in coco format `self._coco_api =COCO(json_file)`.

**Comment by HaoZhang534** [comment]
Hi @weicheng113, I suggest you visualize the predictions on the validation set to see wether the problem is on the model or the evaluator.

**Comment by weicheng113** [comment]
Hi @HaoZhang534 , thanks a lot. I got it working by using MeanAveragePrecision from [torchmetrics](https://torchmetrics.readthedocs.io/en/stable/detection/mean_average_precision.html), which has simpler interface. I will continue with training. When I get more metrics information, I will ask for your advice with finetuning.

**Comment by weicheng113** [comment]
@HaoZhang534 may I ask for your suggestion.

I have a dataset about 1,100 examples, 20% for validation and 80% for training. The dataset only has two categories, 0 and 1.

Below is the histogram of number of instances in each image file. Most of them has two objects in an image and maximum 4 objects in an image.

![image](https://user-images.githubusercontent.com/1066604/207740306-9ab9361a-5489-40f9-a607-d3e86b4608fe.png)

Category 1 has more less instances than category 0 as shown below, which I think it is ok for focal loss.
![image](https://user-images.githubusercontent.com/1066604/207740667-d852a759-17d0-44a5-a546-e3f32bf3af40.png)

=========

I tried with the following configuration.
```python
num_queries: 40
num_dn_queries: 10(if there is 2 instances in an image, it will have 5 DN groups - 5* [2 * 2])
num_select: 20 ( select top 20 confident prediction for validation)
num_epoch: 50 ( with StepLR(optimizer=optimizer, step_size=20) )
train_batch_size: 2
gradient_accumulation_steps: 1 (only on single gpu machine. doing gradient decent at every step).
```

I got below, which is not as good as our yolo version:
```python
{'map': tensor(0.7773),
 'map_50': tensor(0.9670),
 'map_75': tensor(0.9138),
 'map_large': tensor(0.8001),
 'map_medium': tensor(0.4979),
 'map_per_class': tensor(-1.),
 'map_small': tensor(-1.),
 'mar_1': tensor(0.5967),
 'mar_10': tensor(0.8588),
 'mar_100': tensor(0.8595),
 'mar_100_per_class': tensor(-1.),
 'mar_large': tensor(0.8821),
 'mar_medium': tensor(0.5853),
 'mar_small': tensor(-1.)}
```

=======
After looking the above instance stats, I am thinking having experiments with the following two configuratiion:

```python
num_queries: 60
num_dn_queries: 16(if there is 2 instances in an image, it will have 8 DN groups - 8* [2 * 2])
num_select: 6 ( instead of using 4, leaving a bit room)
num_epoch: 50 ( with StepLR(optimizer=optimizer, step_size=20) )
train_batch_size: 2
gradient_accumulation_steps: 1 (only on single gpu machine. doing gradient decent at every step).
```

```python
num_queries: 60
num_dn_queries: 16(if there is 2 instances in an image, it will have 8 DN groups - 8* [2 * 2])
num_select: 6 ( instead of using 4, leaving a bit room)
num_epoch: 50 ( with StepLR(optimizer=optimizer, step_size=20) )
train_batch_size: 2
gradient_accumulation_steps: 3 (only on single gpu machine. doing gradient decent at every 3 steps).
```

Any advice is highly appreciated.

Thanks, Cheng

**Comment by HaoZhang534** [comment]
Is the mAP 0.777% or 77.7%? Given that you have at most 4 instances per image. You can use small num_queries such as 40. num_select can be set larger such as 40. You can also load our coco pre-trained model to fine-tune the model on your dataset.

**Comment by weicheng113** [comment]
@HaoZhang534 Thanks a lot. It is mAP 77.7%. 

Could you explain why to use larger num_select? I thought num_select should be close to max number of instances?

Good suggestion. I will try with pre-trained model out.

Thanks, Cheng

**Comment by HaoZhang534** [comment]
@weicheng113 Larger num_select usually leads to higher AP. About category 1, you may try some tricks to balance the proportions of the two categories. For example, you can add some copies of images from category 1 in the training data.

**Comment by weicheng113** [comment]
@HaoZhang534 Thanks. For final inference, is there a rule to pick the confidence_threshold? I saw 0.5 as default in detrex and 0.3 in DINO repo.

**Comment by HaoZhang534** [comment]
@weicheng113 Which confidence_threshold do you mean? We do not use confidence_threshold to evaluate. We use confidence_threshold=0.3 for visualization.

**Comment by weicheng113** [comment]
Got you, thanks a lot @HaoZhang534 .

**Comment by weicheng113** [comment]
@HaoZhang534 I realized I can't directly load a pretrained dino model, as there is an incompatibility in num_classes, num_queries and num_dn_queries. I can load other weights like transformer weights, but something like class_embed, label_enc need to be retrained.

I will have a try to see if loading a pretrained dino model is helpful. Thanks.

**Comment by HaoZhang534** [comment]
@weicheng113 You are welcome. Your concern is reasonable. It's really a problem when objects are crowded. Maybe some improvements can be made to fix this such as only use negative examples when objects are not crowded.

**Comment by ma3252788** [comment]
Excuse me, can you tell me how to load the custom dataset?

Would be very grateful if it could be done.

**Comment by jiachen0212** [comment]
> Hi @HaoZhang534 , thanks a lot. I got it working by using MeanAveragePrecision from [torchmetrics](https://torchmetrics.readthedocs.io/en/stable/detection/mean_average_precision.html), which has simpler interface. I will continue with training. When I get more metrics information, I will ask for your advice with finetuning.

halou i got the same question，can you give more detail about how to fix this bug？
